# ДЗ 3

In [1]:
import pandas as pd

## Задача 1
```
Вывести на экран первые 5 строк
1.2 Создать новый признак delta_renovated, который будет содержать разницу в годах между годом реновацией дома и годом постройки дома
Если реновации дома не было, то в новом признаке поставьте 0
1.3 Создайте признаки года продажи, месяца продажи
1.4 Удалите признаки date, zipcode, lat, long
```

In [2]:
df = pd.read_csv("lesson2/kc_house_data.csv", sep=",")
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

### 1.2

In [3]:
df['delta_renovated'] = 0
df.loc[df['yr_renovated'] != 0, 'delta_renovated'] = df['yr_renovated'] - df['yr_built']

df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...        1180              0   
1      7242     2.0           0     0  ...        2170            400   
2     10000     1.0           0     0  ...         770              0   
3      5000     1.0           0     0  ...        1050            910   
4      8080     1.0           0     0  ...        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  delta_renovated  
0        5650                0  
1        7639               40  
2        8062                0  
3        5000                0  
4        7503                0  

[5 rows x 22 columns]

### 1.3

In [4]:
df['yr_sale'] = df['date'].apply(
    lambda dt: int(dt[:4])
)

In [5]:
df['mon_sale'] = df['date'].apply(
    lambda dt: int(dt[4:6])
)

In [6]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  yr_built  yr_renovated  zipcode  \
0      5650     1.0           0     0  ...      1955             0    98178   
1      7242     2.0           0     0  ...      1951          1991    98125   
2     10000     1.0           0     0  ...      1933             0    98028   
3      5000     1.0           0     0  ...      1965             0    98136   
4      8080     1.0           0     0  ...      1987             0    98074   

       lat     long  sqft_living15  sqft_lot15  delta_renovated  yr_sale  \
0  47.5112 -122.257           1340        5650                0     2014   
1  47.7210 -122.319           1690        7639               40     2014   
2  47.7379 -122.233           2720        8062                0     2015   
3  47.5208 -122.393           1360        5000                0     2014   
4  47.6168 -122.045           1800        7503                0     2015   

   mon_sale  
0        10  
1        12  
2         2  
3        12  
4         2  

[5 rows x 24 columns]

### 1.4

In [7]:
df.drop(columns=['date', 'zipcode', 'lat', 'long'], inplace=True)
df.head()

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  \
0  7129300520  221900.0         3       1.00         1180      5650     1.0   
1  6414100192  538000.0         3       2.25         2570      7242     2.0   
2  5631500400  180000.0         2       1.00          770     10000     1.0   
3  2487200875  604000.0         4       3.00         1960      5000     1.0   
4  1954400510  510000.0         3       2.00         1680      8080     1.0   

   waterfront  view  condition  grade  sqft_above  sqft_basement  yr_built  \
0           0     0          3      7        1180              0      1955   
1           0     0          3      7        2170            400      1951   
2           0     0          3      6         770              0      1933   
3           0     0          5      7        1050            910      1965   
4           0     0          3      8        1680              0      1987   

   yr_renovated  sqft_living15  sqft_lot15  delta_renovated  yr_sale  mon_sale  
0             0           1340        5650                0     2014        10  
1          1991           1690        7639               40     2014        12  
2             0           2720        8062                0     2015         2  
3             0           1360        5000                0     2014        12  
4             0           1800        7503                0     2015         2

## Задача 2
```
Создайте датафрейм с клиентами:
clients = pd.DataFrame({
'client_id': [1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474],
'house_id': [8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
5083000375, 7920100025, 1951600150, 809001400, 339600110,
1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
8861700030, 3303980210, 7731100066, 8146100580, 825069097,
3889100029, 9524100196]
})
2.1 Присоедините к таблице clients данные по домам через метод join
2.2 Присоедините к таблице clients данные по домам через метод merge
Это нужно, чтобы понимать, какие дома покупались клиентами
house_id - это индексы датафрейма с домами
```

In [8]:
clients = pd.DataFrame({
    'client_id': [
        1459, 4684, 3498, 3942, 4535, 2308, 2866, 2765, 1472, 4236, 2295,
        939, 3840, 280, 20, 4332, 3475, 4213, 3113, 4809, 2134, 2242,
        2068, 4929, 1384, 1589, 3317, 2260, 1727, 1764, 1611, 1474
    ],
    'house_id': [
        8965450190, 6823100225, 5104540330, 2131701075, 1522700060,
        1189000207, 6821600300, 7137950720, 9510920050, 6131600255,
        5428000070, 1788800910, 8100400160, 3123049142, 6306800010,
        5083000375, 7920100025, 1951600150, 809001400, 339600110,
        1622049154, 1099600250, 8563000110, 2768100205, 3995700435,
        8861700030, 3303980210, 7731100066, 8146100580, 825069097,
        3889100029, 9524100196
    ]
}).set_index('house_id')

In [9]:
clients.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32 entries, 8965450190 to 9524100196
Data columns (total 1 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   client_id  32 non-null     int64
dtypes: int64(1)
memory usage: 512.0 bytes


### 2.1

In [10]:
clients.join(df.set_index('id')).head()

client_id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
339600110        4809  395000.0         3       2.50         1610      3755   
809001400        3113  925000.0         3       1.00         1630      3200   
825069097        1764  770000.0         3       2.50         2650     40705   
1099600250       2242  260000.0         3       1.75         1710      6400   
1189000207       2308  387000.0         2       2.50         1170      1394   

            floors  waterfront  view  condition  grade  sqft_above  \
339600110      2.0           0     0          3      7        1610   
809001400      1.5           0     0          4      8        1630   
825069097      2.0           0     0          3      9        2650   
1099600250     1.0           0     0          4      7        1240   
1189000207     2.0           0     0          3      8        1170   

            sqft_basement  yr_built  yr_renovated  sqft_living15  sqft_lot15  \
339600110               0      1987             0           1300        3823   
809001400               0      1912             0           1710        3600   
825069097               0      1994             0           2550       42625   
1099600250            470      1976             0           1600        6400   
1189000207              0      2001             0           1250        3136   

            delta_renovated  yr_sale  mon_sale  
339600110                 0     2014         9  
809001400                 0     2014         9  
825069097                 0     2014         6  
1099600250                0     2014        12  
1189000207                0     2014        10

### 2.2

In [11]:
clients.merge(df, left_on='house_id', right_on='id').info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 33 entries, 0 to 32
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   client_id        33 non-null     int64  
 1   id               33 non-null     int64  
 2   price            33 non-null     float64
 3   bedrooms         33 non-null     int64  
 4   bathrooms        33 non-null     float64
 5   sqft_living      33 non-null     int64  
 6   sqft_lot         33 non-null     int64  
 7   floors           33 non-null     float64
 8   waterfront       33 non-null     int64  
 9   view             33 non-null     int64  
 10  condition        33 non-null     int64  
 11  grade            33 non-null     int64  
 12  sqft_above       33 non-null     int64  
 13  sqft_basement    33 non-null     int64  
 14  yr_built         33 non-null     int64  
 15  yr_renovated     33 non-null     int64  
 16  sqft_living15    33 non-null     int64  
 17  sqft_lot15       3

При этом, поскольку в df имеются дупликаты по id, объединённые данные имеют на 1 строку больше, чем в clients.

Полный список дипликатов id с кол. копий:

In [12]:
inds = df['id'].value_counts()
dupls = inds.loc[inds > 1]
dupls.sort_index()

1000102       2
7200179       2
109200390     2
123039336     2
251300110     2
             ..
9407110710    2
9809000020    2
9828200460    2
9834200305    2
9834200885    2
Name: id, Length: 176, dtype: int64

In [13]:
df[df['id'].isin(dupls.index)].sort_values(by='id')

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
2497      1000102  300000.0         6       3.00         2400      9373   
2496      1000102  280000.0         6       3.00         2400      9373   
16814     7200179  150000.0         2       1.00          840     12750   
16815     7200179  175000.0         2       1.00          840     12750   
11434   109200390  250000.0         3       1.75         1480      3900   
...           ...       ...       ...        ...          ...       ...   
6345   9828200460  260000.0         2       1.00          700      4800   
15200  9834200305  615000.0         3       1.00         1790      3876   
15199  9834200305  350000.0         3       1.00         1790      3876   
1085   9834200885  360000.0         4       2.50         2080      4080   
1086   9834200885  550000.0         4       2.50         2080      4080   

       floors  waterfront  view  condition  grade  sqft_above  sqft_basement  \
2497      2.0           0     0          3      7        2400              0   
2496      2.0           0     0          3      7        2400              0   
16814     1.0           0     0          3      6         840              0   
16815     1.0           0     0          3      6         840              0   
11434     1.0           0     0          4      7        1480              0   
...       ...         ...   ...        ...    ...         ...            ...   
6345      1.0           0     0          3      7         700              0   
15200     1.5           0     0          5      7        1090            700   
15199     1.5           0     0          5      7        1090            700   
1085      1.0           0     0          5      7        1040           1040   
1086      1.0           0     0          5      7        1040           1040   

       yr_built  yr_renovated  sqft_living15  sqft_lot15  delta_renovated  \
2497       1991             0           2060        7316                0   
2496       1991             0           2060        7316                0   
16814      1925             0           1480        6969                0   
16815      1925             0           1480        6969                0   
11434      1980             0           1830        6956                0   
...         ...           ...            ...         ...              ...   
6345       1922             0           1440        4800                0   
15200      1904             0           1360        4080                0   
15199      1904             0           1360        4080                0   
1085       1962             0           1340        4080                0   
1086       1962             0           1340        4080                0   

       yr_sale  mon_sale  
2497      2015         4  
2496      2014         9  
16814     2014        10  
16815     2015         4  
11434     2014        10  
...        ...       ...  
6345      2014         6  
15200     2015         2  
15199     2014         7  
1085      2014         7  
1086      2015         4  

[353 rows x 20 columns]

Другой способ:

In [14]:
df[df.duplicated('id', keep=False)].sort_values(by='id')

id     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
2497      1000102  300000.0         6       3.00         2400      9373   
2496      1000102  280000.0         6       3.00         2400      9373   
16814     7200179  150000.0         2       1.00          840     12750   
16815     7200179  175000.0         2       1.00          840     12750   
11434   109200390  250000.0         3       1.75         1480      3900   
...           ...       ...       ...        ...          ...       ...   
6345   9828200460  260000.0         2       1.00          700      4800   
15200  9834200305  615000.0         3       1.00         1790      3876   
15199  9834200305  350000.0         3       1.00         1790      3876   
1085   9834200885  360000.0         4       2.50         2080      4080   
1086   9834200885  550000.0         4       2.50         2080      4080   

       floors  waterfront  view  condition  grade  sqft_above  sqft_basement  \
2497      2.0           0     0          3      7        2400              0   
2496      2.0           0     0          3      7        2400              0   
16814     1.0           0     0          3      6         840              0   
16815     1.0           0     0          3      6         840              0   
11434     1.0           0     0          4      7        1480              0   
...       ...         ...   ...        ...    ...         ...            ...   
6345      1.0           0     0          3      7         700              0   
15200     1.5           0     0          5      7        1090            700   
15199     1.5           0     0          5      7        1090            700   
1085      1.0           0     0          5      7        1040           1040   
1086      1.0           0     0          5      7        1040           1040   

       yr_built  yr_renovated  sqft_living15  sqft_lot15  delta_renovated  \
2497       1991             0           2060        7316                0   
2496       1991             0           2060        7316                0   
16814      1925             0           1480        6969                0   
16815      1925             0           1480        6969                0   
11434      1980             0           1830        6956                0   
...         ...           ...            ...         ...              ...   
6345       1922             0           1440        4800                0   
15200      1904             0           1360        4080                0   
15199      1904             0           1360        4080                0   
1085       1962             0           1340        4080                0   
1086       1962             0           1340        4080                0   

       yr_sale  mon_sale  
2497      2015         4  
2496      2014         9  
16814     2014        10  
16815     2015         4  
11434     2014        10  
...        ...       ...  
6345      2014         6  
15200     2015         2  
15199     2014         7  
1085      2014         7  
1086      2015         4  

[353 rows x 20 columns]

## Задача 3
```
Составьте несколько сводных таблиц
3.1 Найдите среднюю стоимость домов в зависимости от количества спален
Отсортируйте от меньшей стоимости к большей
3.2 Найдите минимальную, среднюю и максимальную стоимости домов в зависимости от состояния дома
3.3 Постройте таблицу с подсчетом количества домов в данных в зависимости от вида на набережную и оценкой вида
3.4 Каких домов в зависимости от этажности и количества спален больше?
3.5 Постройте таблицу с подсчетом медианной стоимости домов в данных в зависимости от состояния дома и оценки дома
```

In [15]:
import re

def numFormat(col):
    return col.apply(
        lambda val:
            f"{val:_.2f}".replace("_", " ")
            if re.fullmatch(r"((\-\d+)|(\d*))(\.\d+)?", str(val))
            else val
    )

### 3.1

In [16]:
df.groupby('bedrooms') \
    .agg({'price': ['mean']}) \
    .sort_values(by=('price', 'mean')) \
    .apply(numFormat) # при указании списка в {'price': ['mean']} в выходной форме выводится операция

price
                  mean
bedrooms              
1           317 642.88
2           401 372.68
0           409 503.85
3           466 232.08
11          520 000.00
4           635 419.50
33          640 000.00
5           786 599.83
10          819 333.33
6           825 520.64
9           893 999.83
7           951 184.66
8         1 105 076.92

### 3.2

In [17]:
df.pivot_table(
    index='condition',
    values='price',
    aggfunc=['min', 'mean', 'max']
).apply(numFormat)

min        mean           max
                price       price         price
condition                                      
1           78 000.00  334 431.67  1 500 000.00
2           80 000.00  327 287.15  2 555 000.00
3           75 000.00  542 012.58  7 062 500.00
4           89 000.00  521 200.39  7 700 000.00
5          110 000.00  612 418.09  3 650 000.00

### 3.3

In [18]:
pd.crosstab(
    index=df['waterfront'],
    columns=df['view']
)

view            0    1    2    3    4
waterfront                           
0           19489  331  955  491  184
1               0    1    8   19  135

### 3.4

In [19]:
ct = pd.crosstab(
    index=df['floors'],
    columns=df['bedrooms'],
    margins=True
) \
.sort_values(by='All', ascending=False) \
.sort_values(by='All', ascending=False, axis=1)
ct

bedrooms    All     3     4     2     5    6    1   7   0   8  9  10  11  33
floors                                                                      
All       21613  9824  6882  2760  1601  272  199  38  13  13  6   3   1   1
1.0       10680  5455  2383  1951   605  104  162   9   4   5  0   1   0   1
2.0        8241  3118  3682   497   775  119   12  19   6   6  4   2   1   0
1.5        1910   786   698   182   185   30   21   7   0   1  0   0   0   0
3.0         613   405    61   123    13    5    3   1   2   0  0   0   0   0
2.5         161    56    58     5    23   14    1   2   0   0  2   0   0   0
3.5           8     4     0     2     0    0    0   0   1   1  0   0   0   0

Больше всего одноэтажных домов (10680). Также больше всего домов в 3-мя спальнями (9824).

Больше всего домов из сочетаний этажи-спальни:

In [20]:
maxVar = ct.iloc[1:, 1:].max().max()

f"Этажей: {ct.index[ct.eq(maxVar).any(axis=1)][0]}, " + \
f"спален: {ct.columns[ct.eq(maxVar).any(axis=0)][0]}, " + \
f"домов: {maxVar}"

'Этажей: 1.0, спален: 3, домов: 5455'

### 3.5

In [21]:
df.pivot_table(
    index='grade',
    columns='condition',
    values='price',
    aggfunc='median',
    fill_value="---"
).apply(numFormat)

condition           1             2             3             4             5
grade                                                                        
1          142 000.00           ---           ---           ---           ---
3                 ---    280 000.00     75 000.00           ---    262 000.00
4          150 000.00    145 000.00    205 000.00    238 525.00           ---
5          190 000.00    180 000.00    234 475.00    229 975.00    227 450.00
6          255 000.00    235 000.00    265 000.00    295 000.00    285 475.00
7          403 500.00    305 000.00    357 500.00    390 000.00    456 000.00
8          932 500.00    429 000.00    485 000.00    571 250.00    696 000.00
9                 ---    715 000.00    689 000.00    823 500.00  1 078 000.00
10                ---  1 752 500.00    890 000.00  1 030 000.00  1 650 000.00
11                ---           ---  1 209 500.00  1 685 000.00  2 050 000.00
12                ---           ---  1 807 500.00  2 125 000.00  1 990 000.00
13                ---           ---  2 888 000.00  5 750 000.00           ---